### ISSUE TITLE
Multiple HTML files generated for the same in features folder

### ISSUE THREAD

**Issue by ali-himindz** [wswebcreation/protractor-multiple-cucumber-html-reporter-plugin#13]
Hi,
Below is the snippet from my config file. Apparently the plugin is generating multiple html files under report/features. Although only the last one is referenced in index.html but under feature I see more than one html output for the same feature. It appears as if the report is being generated after every feature is completed. 
7f6ed382-5e4f-4a51-a4c6-e39ef56a7378-38255--feature-1.html
7f9f34f8-c696-4e92-9846-be80dfea7d1b-38255--feature-1.html

```
plugins: [{
    package: 'protractor-multiple-cucumber-html-reporter-plugin',
    options: {
      automaticallyGenerateReport: true,
      removeExistingJsonReportFile: true,
      removeOriginalJsonReportFile: true,
      saveCollectedJSON: true,
      displayDuration: true,
      customMetadata: true, 
    }
  }],
````
```config.multiCapabilities = [
  {
    browserName: 'chrome',
    shardTestFiles: true,
    maxInstances: 2,
    specs: [
      '../../../at/regression/*.feature'
    ],
    chromeOptions: {
      args: ['disable-infobars', '--start-maximized']
    },
    metadata: {
      browser: {
        name: 'chrome',
        version: '58'
      }
    }
  }```

**Comment by wswebcreation** [comment]
Hi @ali-himindz 

When I look at you description I wonder how many JSON files you have before the tool will generate the report. It looks like the generated JSON file after each feature is overwritten each time.

Can you give me:

- your complete config file
- the protractor / cucumberjs / protractor-cucumber-framework versions your are using
- check how many generated json file you get, so the one in the json-output folder and how many you would expect

Tnx in advance

**Comment by jayrepo** [comment]
@wswebcreation 
Same issue:
The feature htmls in `report/features` of previous executions is not cleared

```js
exports.config = {

  seleniumAddress: 'http://localhost:4444/wd/hub',
  baseUrl: 'http://juliemr.github.io/protractor-demo',
  directConnect: true,
  restartBrowserBetweenTests: false,
  capabilities: {
    browserName: 'chrome',
    chromeOptions: {
      args: ['disable-infobars']
    }
  },
  framework: 'custom',
  frameworkPath: require.resolve('protractor-cucumber-framework'),
  specs: [
    './features/*.feature'
  ],
  // cucumber command line options
  cucumberOpts: {
    format: 'json:reports/results.json',
    tags: 'not @add'
  },
  plugins: [{
    package: 'protractor-multiple-cucumber-html-reporter-plugin',
    options: {
      automaticallyGenerateReport: true,
      removeExistingJsonReportFile: true
    }
  }]
}
```
```json
  "dependencies": {
    "chai": "^4.1.2",
    "cucumber": "^4.1.0",
    "dirty-chai": "^2.0.1",
    "protractor": "^5.3.0",
    "protractor-cucumber-framework": "^4.2.0",
    "protractor-multiple-cucumber-html-reporter-plugin": "^1.6.1"
  }
```

**Comment by jayrepo** [comment]
@wswebcreation 
1. The number of htmls in `report\features` is keep growing. It does not clear previous results. Is it by design?
2. I have 2 features, with 2 capabilities, it generates 6 htmls in that folder, should it be 4?

I have the demo project at: https://github.com/jayrepo/protractor-cucumber-quick-start

**Comment by wswebcreation** [comment]
> The number of htmls in report\features is keep growing. It does not clear previous results. Is it by design?

Yes, If you want to clear the report you need to doo it yourself with a simple command. The report itself should be a s _dumb_ as possible

> I have 2 features, with 2 capabilities, it generates 6 htmls in that folder, should it be 4?

It should be 4, I don't have time to checkout your project now, maybe next week so I can check it

> I have the demo project at: https://github.com/jayrepo/protractor-cucumber-quick-start

Can you set `removeExistingJsonReportFile` and `removeOriginalJsonReportFile` to false? and place the result in a gist so I can see it?

**Comment by jayrepo** [comment]
> Can you set removeExistingJsonReportFile and removeOriginalJsonReportFile to false? and place the result in a gist so I can see it?

I pushed all the jsons and generated reports to a temp branch:
https://github.com/jayrepo/protractor-cucumber-quick-start/tree/temp/reports/

**Comment by wswebcreation** [comment]
Ahh, tnx for the project and the files. I now know what you mean and I can explain why this is happening.

First of all I'm limited by the plugin architecture of Protractor, see [`teardown`](https://github.com/angular/protractor/blob/master/lib/plugins.ts#L72), [`postResults`](https://github.com/angular/protractor/blob/master/lib/plugins.ts#L87), [`postTests`](https://github.com/angular/protractor/blob/master/lib/plugins.ts#L106). These are the hooks I can use to generate the report. But because the plugin architecture only hooks into a _single_ instance and not waiting for all instances to be finished I can only generate a report after each instance that has been done and released. 

Resulting in for example 

- Chrome being faster than firefox => 1 report => 2 files
- Firefox done, search the report folder, seeing chrome (2 files) already done, generate report for Firefox (2 files) and Chrome (2 files) => 4 result in your newly generated overview, 6 files in your features folder which hold the already generate feature html pages from chrome.

When I created this plugin I also thought about removing the old files, but because of parallel executions, where you never know the right order when tests are done, I didn't dare to remove the previous feature html pages.

I hope I explained it and you understand it. If you have a better way, please let me know. By the way, I also thought about removing the report generation out of the plugin, but then you need to write it in your own post hook and I wanted to make the usage of the plugin as simple as possible.

I'll update the docs and link to this answer. I'm going to close it for know, but it's still open for discussion.

**Comment by jayrepo** [comment]
@wswebcreation Understood, thanks for your explanation. I've tried many cucumber reporters, this one is the easiest to use.

For removing old files
One possible solution I can think of is to get the latest json's modified date when each instance is done, then remove all feature html pages of which created date is earlier than that modified date.
Or if we don't need to see the partial results while some instances is still running, simply empty that folder when each instance is done, as the slowest instance will regenerate all the htmls based on json.

**Comment by wswebcreation** [comment]
Hi @jayrepo 

Tnx for the compliment.

I've also been thinking about removing the files, but there could a situation that you remove the folder, but the last instance is also ready which removes the complete report resulting in having nothing.

**Comment by jayrepo** [comment]
Hi @wswebcreation 

Not remove the folder, do sth like `rm -rf *.html`  in `report\features` folder before generate new htmls. The worst situation should be duplicated results if all instances completed at the same time.

Or a safer way is to get the modified date of all jsons, find the earliest one. Then delete htmls created earlier than that date.

**Comment by ebacka** [comment]
Hello,

What about some more details like line, branch, functions coverage??? Is it possible in this plugin?
